### ISSUE TITLE
Compiling issues

### ISSUE THREAD

**Issue by mbelouso** [czimaginginstitute/MotionCor3#11]
Hey guys,

I have the following issue. When compiling with:

`make exe -f makefile11`

I get the following error after it compiles lots of other things:

`/usr/bin/ld: /home/mbelouso/MotionCor3/Lib/libmrcfile.a(CLoadExtHeader.o): relocation R_X86_64_32 against symbol '__gxx_personality_v0@@CXXABI_1.3' can not be used when making a PIE object; recompile with -fPIE
/usr/bin/ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
make: *** [makefile11:160: exe] Error 1`

gcc and g++ version is 11.4, CUDA 11.8, linux mint 21

Any thoughts?

**Comment by elainehoml** [comment]
I had the same issue but adding `-Xcompiler -no-pie` to makefile11 in the exe block worked for me (copied from Issue #4)

See https://github.com/czimaginginstitute/MotionCor3/issues/4#issuecomment-1784848759 for the full code snippet

**Comment by mbelouso** [comment]
@elainehoml Thanks for the heads up. I added these lines to the makefile11, but still run into the same issue. Same error output as in my initial post.

**Comment by stiobhan** [comment]
Adding these flags in makefile11 seems to work. A different workaround that worked for me as well was to use GCC 10 instead of GCC 11.

**Comment by mbelouso** [comment]
Yeah. I eventually got it to compile with a different version of GCC (8.0)
and adding these flags. Thanks.

**Comment by jpellman** [comment]
Just to provide more context here, at some point several popular Linux distributions decided to compile their packaged versions of gcc with the `-enable-default-pie` option (see [here](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=103398)).  The version of gcc used isn't as relevant so much as whether or not gcc was compiled with PIE as default.  The list of compile options for the particular gcc binary you're using can be found by typing `gcc -v`.

A [PIE (position independent executable)](https://en.wikipedia.org/wiki/Position-independent_code#Position-independent_executables) is a binary compiled with security measure that is meant to prevent some common buffer overflow attacks; I'm not certain, but I imagine it has become the default due to several high-profile buffer overflow exploits in somewhat recent years (e.g., [Heartbleed](https://en.wikipedia.org/wiki/Heartbleed)).

I'm uncertain why PIE binaries can't be produced with NVCC, but some [frightfully old CUDA docs](https://developer.download.nvidia.com/compute/cuda/4_1/rel/toolkit/docs/CUDA_Toolkit_Release_Notes_And_Errata.txt) provide a clue:

```
The host linker on Mac OS 10.7 generates position-independent executables by default. As CUDA does not support position-independent executable currently, the linker must generate position-dependent executable by passing in the -no_pie option. If nvcc is being used to link the application, this option will be passed to the linker by default. To override the default behavior, the -Xlinker -pie option can be passed to nvcc.
```

My educated guess is that modern versions of CUDA _still_ do not support PIE binaries, although I can find no documentation out there to confirm or deny this hypothesis.
